concordium-consensus/src/Concordium/Queries.hs

```diff
@@ -1207,6 +1207,136 @@ getDelegatorsRewardPeriod bhi maybeBakerId = do
               pdrpiStake = dcDelegatorCapital
             }
 
+-- ** Epoch indexed
+
+data EpochQueryError
+    = -- |The specified block was not found.
+      EQEBlockNotFound
+    | -- |The specified epoch or genesis index is in the future.
+      EQEFutureEpoch
+    | -- |The epoch is not valid for the specified genesis index.
+      EQEInvalidEpoch
+    | -- |The genesis index does not support the query.
+      EQEInvalidGenesisIndex
+
+-- |Get the first finalized block in a specified epoch.
+-- The epoch can be specified directly by the genesis index and epoch number, or indirectly by a
+-- 'BlockHashInput' that references a block.
+--
+-- The following failure cases apply:
+--   * If the input is a 'BlockHashInput' and the specified block cannot be resolved, this returns
+--     'EQEBlockNotFound'.
+--   * If the specified genesis index is greater than the current one, returns 'EQEFutureEpoch'.
+--   * If the specified epoch is after the epoch of the last finalized block
+--      - returns 'EQEFutureEpoch' if the genesis index is the current one
+--      - returns 'EQEInvalidEpoch' otherwise.
+--   * If the specified epoch contains no finalized blocks then returns 'EQEBlockNotFound'.
+--     This only applies to consensus version 0, as epochs cannot be empty in consensus version 1.
+getFirstBlockEpoch :: forall finconf. EpochRequest -> MVR finconf (Either EpochQueryError BlockHash)
+getFirstBlockEpoch SpecifiedEpoch{..} = MVR $ \mvr -> do
+    versions <- readIORef $ mvVersions mvr
+    case versions Vec.!? fromIntegral erGenesisIndex of
+        Nothing -> return $ Left EQEFutureEpoch
+        Just evc -> do
+            let isCurrentVersion = fromIntegral erGenesisIndex == Vec.length versions
+            liftSkovQuery
+                mvr
+                evc
+                ( do
+                    -- Consensus version 0
+                    getFirstFinalizedOfEpoch (Left erEpoch) <&> \case
+                        Left FutureEpoch
+                            | isCurrentVersion -> Left EQEFutureEpoch
+                            | otherwise -> Left EQEInvalidEpoch
+                        Left EmptyEpoch -> Left EQEBlockNotFound
+                        Right block -> Right $! getHash @BlockHash block
+                )
+                ( do
+                    -- Consensus version 1#
+                    (SkovV1.getFirstFinalizedBlockOfEpoch (Left erEpoch) =<< get) <&> \case
+                        Nothing
+                            | isCurrentVersion -> Left EQEFutureEpoch
+                            | otherwise -> Left EQEInvalidEpoch
```

Comment:
The functions looking up the block for each tree state looks very similar, but yet there is a discrepancy with the result codes, i.e. for consensus version 0 there is a `EQEBlockNotFound`, but not for the latter? By looking at the code for consensus version 1 it looks like the two cases will be treated the same (`EQEFutureEpoch`) ?

Reply:
Also, couldnt the two consensus versions actually share the underlying implementation by passing in the handles to the respective treestate stores? 

I guess my question is, why are two separate implementations required for something that is essentially the same except for the treestate store handles (and the bp type).

Reply:
The `EQEBlockNotFound` case cannot happen in consensus version 1 because it requires an epoch to be empty, which is (theoretically) possible in consensus version 0.

The implementations do share `interpolationSearchFirstM`, which is the bulk of the look-up. I don't think unifying the implementations beyond this is useful, as there is no common interface over the treestate for the different consensus versions.